Transform improvements

maZahaca opened the pull request:
- Moved transforms by different files
- Added missing lanGoose definitions for transforms

:goose:

maZahaca commented:
@jifeon review it please

jifeon commented on lib/Parser.js:
```diff
@@ -290,15 +289,22 @@ Parser.prototype = {
-    addTransformation: function() {
-        return this._transforms.addTransformation.apply(this._transforms, arguments)
+    addTransformation: function(type, transform) {
+        return this.addTransform(type, transform);
```
I would add `debug` message here as well, kinda warning

jifeon commented on lib/Transforms.js:
```diff
@@ -0,0 +1,56 @@
+const Storage = require('./Storage');
+const transformsFactory = require('./transforms/transformsFactory');
+
+function Transforms(options) {
```
you are using es6 classes bellow, why to not use them here?

jifeon commented on lib/transforms/TransformCombine.js:
```diff
@@ -0,0 +1,33 @@
+        const fields = this._options.fields || [];
+        const type = this._options.dataType || 'string';
+        const resultValue = [];
+        _.each(fields, field => {
```
pls use `Array#forEach`

jifeon commented on lib/transforms/TransformJoin.js:
```diff
@@ -0,0 +1,18 @@
+
+class TransformJoin extends Transform {
+    doTransform() {
+        const glue = this._options.glue || ' ';
```
so `glue` cannot be empty string? `'' || ' ' === ' '`

jifeon commented on lib/transforms/TransformMatch.js:
```diff
@@ -0,0 +1,36 @@
+        const re = this._options.re;
+        const index = this._options.index || 0;
+
+        matches = (this._value || '').match(RegExp.apply(null, re));
```
and if `this._value` is zero? I think need to use `String` here

maZahaca replied:
Which options are available to do that?

jifeon commented on lib/transforms/TransformReplace.js:
```diff
@@ -0,0 +1,22 @@
+            throw new Error('You must pass an array as `re` to `replace` transform');
+        }
+
+        return (this._value || '').replace(RegExp.apply(null, re), this._options.to);
```
the same here

jifeon commented on lib/transforms/TransformSplit.js:
```diff
@@ -0,0 +1,27 @@
+        const index = this._options.index || 0;
+        const dataType = this._options.dataType || 'string';
+        let separator = this._options.separator;
+        separator = Array.isArray(separator) ? RegExp.apply(null, separator) : separator || ',';
```
separator cannot be empty string?

maZahaca commented:
@jifeon Can you have a look on [TransformPick](https://github.com/redco/goose-parser/blob/1154c7c066104bb12cc37ffa46d263d31865d407/lib/transforms/TransformPick.js) and [TransformGet](https://github.com/redco/goose-parser/blob/1154c7c066104bb12cc37ffa46d263d31865d407/lib/transforms/TransformGet.js). Seems they are duplicates.

jifeon commented:
@maZahaca is it ready for re-review?

maZahaca commented:
I fixed issues, you mentioned before.
For now I just added one more transform.

I've tried to run it on real data, and some transform fail. Gonna fix that issue
